saya/core/src/block_ingestor/polling.rs

```diff
@@ -1,127 +1,275 @@
+    channel: Option<Sender<BlockInfo>>,
+    block_pie_generator: B,
+    db: DB,
+    workers_count: usize,
 }
 
-impl<S> PollingBlockIngestor<S>
+impl<S, B, DB> PollingBlockIngestor<S, B, DB>
 where
-    S: AsRef<[u8]>,
+    S: AsRef<[u8]> + Send + Sync + Clone + 'static,
+    B: BlockPieGenerator + Send + Sync + Clone + 'static,
+    DB: PersistantStorage + Send + Sync + Clone + 'static,
 {
-    async fn run(mut self) {
-        let provider = JsonRpcClient::new(HttpTransport::new(
-            Url::parse(self.rpc_url.as_str()).unwrap(),
-        ));
+    /// Fetches the latest block number from the StarkNet RPC.
+    async fn get_latest_block(&self) -> Option<u64> {
+        let provider = JsonRpcClient::new(HttpTransport::new(self.rpc_url.clone()));
 
-        loop {
-            let latest_block = provider.block_number().await.unwrap();
+        let block_number = crate::utils::retry_with_backoff(
+            || provider.block_number(),
+            "get_latest_block",
+            MAX_RETRIES as u32,
+            Duration::from_secs(5),
+        )
+        .await;
 
-            if self.current_block > latest_block {
-                debug!(
-                    "Current block {} is greater than latest block {}",
-                    self.current_block, latest_block
-                );
+        match block_number {
+            Ok(block_number) => Some(block_number),
+            Err(err) => {
+                error!("Failed to fetch latest block: {}", err);
+                None
+            }
+        }
+    }
 
-                tokio::select! {
-                    _ = self.finish_handle.shutdown_requested() => break,
-                    _ = tokio::time::sleep(PROVE_BLOCK_FAILURE_BACKOFF) => continue,
-                }
+    /// Worker function: proves a block and sends the result.
+    async fn worker(
+        task_rx: Arc<Mutex<mpsc::Receiver<u64>>>,
+        block_pie_generator: B,
+        finish_handle: FinishHandle,
+        rpc_url: Url,
+        channel: mpsc::Sender<BlockInfo>,
+        snos: S,
+        db: DB,
+    ) where
+        S: AsRef<[u8]> + Send + Sync + 'static,
+        B: BlockPieGenerator + Send + Sync + 'static,
+        DB: PersistantStorage + Send + Sync + 'static,
+    {
+        loop {
+            let block_number = if let Some(block_number) = task_rx.lock().await.recv().await {
+                block_number
+            } else {
+                break;
+            };
+
+            if finish_handle.is_shutdown_requested() {
+                break;
             }
 
-            let pie = match prove_block::prove_block(
-                self.snos.as_ref(),
-                self.current_block,
-                // This is because `snos` expects a base URL to be able to derive `pathfinder` RPC path.
-                self.rpc_url.as_str().trim_end_matches("/rpc/v0_7"),
-                LayoutName::all_cairo,
-                true,
+            crate::utils::retry_with_backoff(
```

Comment:
I think for the retry with backoff it's more for network operations.

At the DB level, it may just be configuration of the connection instead.